src/components/Elements/TOC/TOC.tsx

```diff
@@ -0,0 +1,74 @@
+import type { MarkdownHeading } from "astro"
+import { onMount, type Component } from "solid-js"
+
+export type TOCProps = {
+  headings: MarkdownHeading[]
+}
+
+export const TOC: Component<TOCProps> = (props) => {
+  let observer: IntersectionObserver
+  let activeHeadingCount: number = 0
+  let pendingDeactivateHeadingElements: HTMLElement[] = []
+
+  const headings = props.headings
+  const headingElements = headings
+    .filter((heading) => heading.depth === 2 || heading.depth === 3)
+    .map((heading) => (
+      <li class="mb-2 ms-4">
+        <div
+          class={`absolute ${
+            heading.depth === 2 ? "-start-[6px] mt-1.5 size-3" : "-start-[3px] mt-2 size-1.5"
+          } rounded-full border border-border-muted bg-border-muted`}
+        />
+        <a href={`#${heading.slug}`} class="block max-w-[200px] truncate">
+          <span
+            class={`text-mauve-11 ${heading.depth === 2 ? "font-bold" : "font-normal"}`}
+            id={`toc-${heading.text.toLowerCase().replace(/ /g, "-")}`}
+          >
+            {heading.text}
+          </span>
+        </a>
+      </li>
+    ))
+
+  onMount(() => {
+    observer = new IntersectionObserver((entries: IntersectionObserverEntry[]) => {
+      entries.forEach((entry) => {
+        const toc = document.getElementById(`toc-${entry.target.id.toLowerCase().replace(/ /g, "-")}`)
```

Comment:
Nit: 可読性のために不必要な正規表現は避けた方が良いと思われます

```suggestion
        const toc = document.getElementById(`toc-${entry.target.id.toLowerCase().replaceAll(" ", "-")}`)
```